### PR TITLE
Adopt Ruff

### PR DESCRIPTION
**Pull request by AA-Turner**
I've set `target-version = "py312"` as that is what we have in CPython, but we could lower this requirement if wanted.

A

----
📚 Documentation preview 📚: https://cpython-devguide--1372.org.readthedocs.build/

### CODE REVIEW

File: .ruff.toml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+target-version = "py312"
+fix = true
```
Comment (ezio-melotti) [inline .ruff.toml:2]: Does this fix the issues automatically on commit?

Comment (AA-Turner) [reply .ruff.toml:2]: Not on commit, but when you run `ruff check`.

File: .ruff.toml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    "G",     # flake8-logging-format
+    "I",     # isort
+    "ISC",   # flake8-implicit-str-concat
+    "LOG",   # flake8-logging
+    "PERF",  # perflint
+    "PGH",   # pygrep-hooks
+    "PT",    # flake8-pytest-style
+    "TCH",   # flake8-type-checking
+    "UP",    # pyupgrade
+    "W",     # pycodestyle
```
Comment (hugovk) [inline .ruff.toml:25]: We had this for Flake8, let's keep it:
```suggestion
    "W",     # pycodestyle
    "YTT",   # flake8-2020
```

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -1,33 +1,13 @@
...
-      - id: python-check-blanket-noqa
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.5.7
+    hooks:
+      - id: ruff-format
+        name: Run Ruff (format)
+        args: [--check]
+      - id: ruff
+        name: Run Ruff (lint)
+        args: [--exit-non-zero-on-fix]
```
Comment (hugovk) [inline .pre-commit-config.yaml:10]: Ruff's pre-commit docs say:

> When running with `--fix`, Ruff's lint hook should be placed _before_ Ruff's formatter hook, and
_before_ Black, isort, and other formatting tools, as Ruff's fix behavior can output code changes
that require reformatting.

https://github.com/astral-sh/ruff-pre-commit#using-ruff-with-pre-commit

```suggestion
      - id: ruff
        name: Run Ruff (lint)
        args: [--exit-non-zero-on-fix]
      - id: ruff-format
        name: Run Ruff (format)
        args: [--check]
```

**Review by hugovk — COMMENTED** [review]
> I've set `target-version = "py312"` as that is what we have in CPython, but we could lower this requirement if wanted.

I don't mind much, but some others have asked for lower versions before. We don't gain a lot by requiring a higher version, just the `datetime` thing at the moment, which isn't too much bother.

File: .ruff.toml [hunk 4]
```diff
@@ -0,0 +1,34 @@
+target-version = "py312"
```
Comment (AA-Turner) [inline .ruff.toml:1]:
```suggestion
target-version = "py311"
```

File: .ruff.toml (hunk 1 above)
Comment (AA-Turner) [reply .ruff.toml:2]: I won't in this PR to keep it more focussed (I'm also unfamiliar with the service), but perhaps something to consider.

A

**Pull request merged**